stl/inc/xstring

```diff
@@ -2589,19 +2607,26 @@ public:
     }
 
 private:
-    static constexpr void _Start_element_lifetimes(_Elem* const _Ptr, const size_type _Size) {
+    _NODISCARD static _CONSTEXPR20 pointer _Allocate_for_capacity(_Alty& _Al, const size_type _Capacity) {
+        _STL_INTERNAL_CHECK(_Entails_large_string(_Capacity));
+        const pointer _Fancy_ptr = _Al.allocate(_Capacity + 1); // +1 for null terminator
+#if _HAS_CXX20
         // Start element lifetimes to avoid UB. This is a more general mechanism than _String_val::_Activate_SSO_buffer,
         // but likely more impactful to throughput.
-#if _HAS_CXX20
         if (_STD is_constant_evaluated()) {
-            for (size_type _Idx = 0; _Idx < _Size; ++_Idx) {
+            _Elem* const _Ptr = _Unfancy(_Fancy_ptr);
+            for (size_type _Idx = 0; _Idx < _Capacity + 1; ++_Idx) {
                 _STD construct_at(_Ptr + _Idx);
             }
         }
-#else // ^^^ C++20-or-later / pre-C++20 vvv
-        (void) _Ptr;
-        (void) _Size;
 #endif // _HAS_CXX20
+        return _Fancy_ptr;
+    }

```

Comment:
As to `_Allocate_at_least_helper` introduced in #3864, I think we can make `_Allocate_for_capacity` apply `_Allocate_at_least_helper` logic, then most of the rest part in this pr will work without change. The only place that needs extra work will be an allocation in the interaction part with `basic_stringbuf`.

Side note:
In the past we have to "explicitly" take care of the extra null terminator for both size(when copy) and capacity(when allocate). If `_(De)allocate_for_capacity` is applied, only size will need manual `+1` logic.